skbio/alignment/_path.py

```diff
@@ -0,0 +1,446 @@
+
+    def __repr__(self):
+        r"""Return summary of the alignment path."""
+        return (
+            f"<{self.__class__.__name__}, shape: {self.shape}, "
+            f"CIGAR: '{self.to_cigar()}'>"
+        )
+
+    @classonlymethod
+    def from_bits(cls, bits):
+        r"""Create an alignment path from a bit array.
+
+        Parameters
+        ----------
+        bits : array_like of 0's and 1's of shape (n_seqs, n_positions)
+            Array of zeros (character) and ones (gap) which represent the alignment.
+
+        Returns
+        -------
+        PairAlignPath
+            The alignment path of the provided bit array.
+        """
+        # Ensure bits is a 2D array-like of ones and zeros.
+        if not isinstance(bits, np.ndarray):
+            bits = np.atleast_2d(bits)
+        if bits.ndim != 2 or bits.shape[1] == 0:
+            raise ValueError("Input 'bits' must be a non-empty 2D numpy array.")
+        if not (np.logical_or(bits == 0, bits == 1).all()):
+            raise ValueError("Input 'bits' must contain only zeros and ones.")
+
+        ints = bits[0] + (bits[1] << 1)
+        idx = np.append(0, np.where(ints[:-1] != ints[1:])[0] + 1)
+        lens = np.append(idx[1:] - idx[:-1], ints.size - idx[-1])
+        return cls(lens, ints[idx], np.zeros(bits.shape[0], dtype=int))
+
+    def to_bits(self):
+        r"""Unpack states into an array of bits."""
+        if not np.all(np.isin(self.states, [0, 1, 2, 3])):
+            raise ValueError(
+                "For pairwise alignment, `states` must only contain "
+                "zeros, ones, twos, or threes."
+            )
+        return np.stack([self.states & 1, self.states >> 1])
+
+    def to_cigar(self, seqs=None):
+        r"""Get a CIGAR string representing the pairwise alignment path.
+
+        Parameters
+        ----------
+        seqs : list of skbio.Sequence or list of string
+            A pair of sequences to generate CIGAR string. If provided, will
+            distinguish match (``=``) and mismatch (``X``). Otherwise, will uniformly
+            note them as (mis)match (``M``). The first sequence in the list is the
+            query sequence, the second is the reference sequence.
+        """
+        cigar = []
+
+        states = np.squeeze(self.states)
+
+        if seqs is not None:
+            # test if seqs is strings or Sequence object or something else
+            if isinstance(seqs[0], str) and isinstance(seqs[1], str):
+                seq1 = np.array([ord(char) for char in str(seqs[0])])
+                seq2 = np.array([ord(char) for char in str(seqs[1])])
+            elif isinstance(seqs[0], GrammaredSequence) and isinstance(
+                seqs[1], GrammaredSequence
+            ):
+                seq1 = seqs[0]._bytes
+                seq2 = seqs[1]._bytes
+            else:
+                raise ValueError("`seqs` must be of type string or Sequence object.")
+
+            idx1, idx2 = [int(x) for x in self.starts]
```

Comment:
I thought that `starts` is always an integer array, therefore `int` is not necessary?

Reply:
`starts` is defined initially as `self.starts = np.asarray(starts, dtype=np.uint64)`, but when I unpacked `starts` with `idx1, idx2 = self.starts`, `idx1` and `idx2` became `float`s when we add them to `length` in line 335. Using `int` here solves this problem.

For example:
````python
>>> starts = np.array([0, 1], dtype=np.uint64)
>>> lengths = np.array([3, 4], dtype=np.int64)
>>> a, b = starts
>>> c, d = lengths
>>> type(a)
<class 'numpy.uint64'>
>>> type(c)
<class 'numpy.int64'>
>>> a + c
3.0
````

The other solution is to use `np.int64` instead of `np.uint64` when we define `starts` at the beginning of the class. `np.uint64` has strange behavior as I have mentioned elsewhere in this PR.

Reply:
Thanks for checking this! I think that using `np.int64` instead of `np.uint64` would be the best solution.